<a href="https://colab.research.google.com/github/anissapatel/ML4VA/blob/main/ML4VA_Traffic_Patterns.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [32]:
!git clone https://github.com/anissapatel/ML4VA.git

fatal: destination path 'ML4VA' already exists and is not an empty directory.


In [33]:
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV, cross_val_score
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error
from scipy.stats import randint
import requests
import geopandas as gpd
import numpy as np
np.random.seed(42)

In [34]:
# Define the API URL
url = "https://services.arcgis.com/p5v98VHDX9Atv3l7/arcgis/rest/services/CrashData_test/FeatureServer/0/query?where=1%3D1&outFields=*&outSR=4326&f=json"
response = requests.get(url)

In [35]:
# Check if the request was successful
if response.status_code == 200:
    print("Data fetched successfully!")
    # Load the GeoJSON data into a GeoDataFrame
    geo_data = gpd.read_file(response.text)
    # Convert to DataFrame if you don’t need geometry
    df = pd.DataFrame(geo_data.drop(columns='geometry'))
    print("Data successfully converted to DataFrame!")
else:
    print(f"Failed to fetch data. Status code: {response.status_code}")

df.head()

Data fetched successfully!
Data successfully converted to DataFrame!


OBJECTID  DOCUMENT_NBR CRASH_YEAR                  CRASH_DT  \
0         1     163465085       2016 2016-12-06 00:00:00+00:00   
1         2     160725125       2016 2016-01-26 00:00:00+00:00   
2         3     160465260       2016 2016-02-13 00:00:00+00:00   
3         4     161255113       2016 2016-03-26 00:00:00+00:00   
4         5     162005180       2016 2016-07-18 00:00:00+00:00   

  CRASH_MILITARY_TM CRASH_SEVERITY  K_PEOPLE  A_PEOPLE  B_PEOPLE  C_PEOPLE  \
0               700              O         0         0         0         0   
1              1636              O         0         0         0         0   
2              1845              B         0         0         2         0   
3                 1              O         0         0         0         0   
4               732              O         0         0         0         0   

   ...  AREA_TYPE             SYSTEM  VSP                   OWNERSHIP  \
0  ...      Urban    VDOT Interstate    7         1. State Hwy Agency   
1  ...      Urban  NonVDOT secondary    5  3. City or Town Hwy Agency   
2  ...      Rural     VDOT Secondary    1         1. State Hwy Agency   
3  ...      Urban     VDOT Secondary    1         1. State Hwy Agency   
4  ...      Rural       VDOT Primary    6         1. State Hwy Agency   

                             PLAN_DISTRICT MPO_NAME                 RTE_NM  \
0                        Northern Virginia     NOVA       R-VA   IS00095NB   
1                            Hampton Roads     HAMP  S-VA114NP WOODLAND RD   
2                            Northern Neck     None       R-VA066SC00640NB   
3                        Richmond Regional     RICH       R-VA042SC00782SB   
4  Roanoke Valley-Alleghany, West Piedmont     None       R-VA   US00220NB   

   RNS_MP     NODE  OFFSET  
0  158.85     None     NaN  
1    0.50   253154  318.32  
2    2.54  1149141    5.82  
3    1.19     None     NaN  
4   48.61   328737    5.34  

[5 rows x 67 columns]

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 67 columns):
 #   Column                    Non-Null Count  Dtype              
---  ------                    --------------  -----              
 0   OBJECTID                  2000 non-null   int32              
 1   DOCUMENT_NBR              2000 non-null   int32              
 2   CRASH_YEAR                2000 non-null   object             
 3   CRASH_DT                  2000 non-null   datetime64[ms, UTC]
 4   CRASH_MILITARY_TM         2000 non-null   object             
 5   CRASH_SEVERITY            2000 non-null   object             
 6   K_PEOPLE                  2000 non-null   int16              
 7   A_PEOPLE                  2000 non-null   int16              
 8   B_PEOPLE                  2000 non-null   int16              
 9   C_PEOPLE                  2000 non-null   int16              
 10  PERSONS_INJURED           2000 non-null   int16              
 11  PEDESTRIANS_KILLE

In [37]:
df = df[["CRASH_YEAR", "CRASH_DT", "CRASH_SEVERITY","COLLISION_TYPE", "WEATHER_CONDITION",
         "LIGHT_CONDITION","ROADWAY_SURFACE_COND", "RELATION_TO_ROADWAY","ROADWAY_ALIGNMENT",
         "ROADWAY_DEFECT", "ROADWAY_DESCRIPTION","INTERSECTION_TYPE","SCHOOL_ZONE",
         "SPEED_NOTSPEED","INTERSECTION_ANALYSIS", "MAINLINE_YN","NIGHT","VDOT_DISTRICT",
         "AREA_TYPE","RTE_NM"]]
df.head()

CRASH_YEAR                  CRASH_DT CRASH_SEVERITY  \
0       2016 2016-12-06 00:00:00+00:00              O   
1       2016 2016-01-26 00:00:00+00:00              O   
2       2016 2016-02-13 00:00:00+00:00              B   
3       2016 2016-03-26 00:00:00+00:00              O   
4       2016 2016-07-18 00:00:00+00:00              O   

               COLLISION_TYPE                       WEATHER_CONDITION  \
0                 1. Rear End  1. No Adverse Condition (Clear/Cloudy)   
1                    2. Angle  1. No Adverse Condition (Clear/Cloudy)   
2  9. Fixed Object - Off Road  1. No Adverse Condition (Clear/Cloudy)   
3                 1. Rear End                                 5. Rain   
4                 1. Rear End  1. No Adverse Condition (Clear/Cloudy)   

                  LIGHT_CONDITION ROADWAY_SURFACE_COND  \
0      4. Darkness - Road Lighted               1. Dry   
1                     2. Daylight               1. Dry   
2      4. Darkness - Road Lighted               1. Dry   
3  5. Darkness - Road Not Lighted               2. Wet   
4                     2. Daylight               1. Dry   

      RELATION_TO_ROADWAY    ROADWAY_ALIGNMENT ROADWAY_DEFECT  \
0     8. Non-Intersection  1. Straight - Level  1. No Defects   
1    1. Main-Line Roadway  1. Straight - Level  1. No Defects   
2     8. Non-Intersection  1. Straight - Level  1. No Defects   
3     8. Non-Intersection  1. Straight - Level  1. No Defects   
4  9. Within Intersection  3. Grade - Straight  1. No Defects   

                            ROADWAY_DESCRIPTION       INTERSECTION_TYPE  \
0  3. Two-Way, Divided, Positive Median Barrier  1. Not at Intersection   
1                       1. Two-Way, Not Divided  1. Not at Intersection   
2                       1. Two-Way, Not Divided  1. Not at Intersection   
3                       1. Two-Way, Not Divided  1. Not at Intersection   
4       2. Two-Way, Divided, Unprotected Median     3. Three Approaches   

  SCHOOL_ZONE SPEED_NOTSPEED INTERSECTION_ANALYSIS MAINLINE_YN NIGHT  \
0       3. No            Yes      Not Intersection         Yes   Yes   
1       3. No             No      Not Intersection         Yes    No   
2       3. No             No     VDOT Intersection         Yes   Yes   
3       3. No             No      Not Intersection         Yes   Yes   
4       3. No             No     VDOT Intersection         Yes    No   

          VDOT_DISTRICT AREA_TYPE                 RTE_NM  
0  9. Northern Virginia     Urban       R-VA   IS00095NB  
1      5. Hampton Roads     Urban  S-VA114NP WOODLAND RD  
2     6. Fredericksburg     Rural       R-VA066SC00640NB  
3           4. Richmond     Urban       R-VA042SC00782SB  
4              2. Salem     Rural       R-VA   US00220NB

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype              
---  ------                 --------------  -----              
 0   CRASH_YEAR             2000 non-null   object             
 1   CRASH_DT               2000 non-null   datetime64[ms, UTC]
 2   CRASH_SEVERITY         2000 non-null   object             
 3   COLLISION_TYPE         2000 non-null   object             
 4   WEATHER_CONDITION      2000 non-null   object             
 5   LIGHT_CONDITION        2000 non-null   object             
 6   ROADWAY_SURFACE_COND   2000 non-null   object             
 7   RELATION_TO_ROADWAY    2000 non-null   object             
 8   ROADWAY_ALIGNMENT      2000 non-null   object             
 9   ROADWAY_DEFECT         2000 non-null   object             
 10  ROADWAY_DESCRIPTION    2000 non-null   object             
 11  INTERSECTION_TYPE      2000 non-null   object           

In [39]:
# Step 3: Define target and feature columns
target_column = "CRASH_SEVERITY"
categorical_columns = ["COLLISION_TYPE", "WEATHER_CONDITION", "LIGHT_CONDITION",
                       "ROADWAY_SURFACE_COND", "RELATION_TO_ROADWAY", "ROADWAY_ALIGNMENT",
                       "ROADWAY_DEFECT", "ROADWAY_DESCRIPTION", "INTERSECTION_TYPE",
                       "SCHOOL_ZONE", "SPEED_NOTSPEED", "INTERSECTION_ANALYSIS",
                       "MAINLINE_YN", "NIGHT", "VDOT_DISTRICT", "AREA_TYPE", "RTE_NM"]

In [40]:
# Split data into features (X) and target (y)
X = df[categorical_columns]
y = df[target_column]

In [41]:
# Encode the target variable
label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)

In [42]:
# Further split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.2, random_state=42)

In [43]:
# Step 4: Define preprocessing and model pipeline
preprocessor = ColumnTransformer([
    ('cat', OneHotEncoder(sparse_output=False, handle_unknown="ignore"), categorical_columns)
])

In [44]:
# Define the pipeline with preprocessing and model
pipeline = Pipeline([
    ('preprocessor', preprocessor),
    ('regressor', RandomForestRegressor(random_state=42))
])

In [45]:
# Step 5: Initial Model Evaluation using Cross-Validation
cross_val_scores = cross_val_score(pipeline, X_train, y_train, scoring="neg_mean_squared_error", cv=10, error_score='raise')
initial_rmse_scores = np.sqrt(-cross_val_scores)
print("Initial Cross-Validation RMSE Scores:", initial_rmse_scores)
print("Mean RMSE:", initial_rmse_scores.mean())

Initial Cross-Validation RMSE Scores: [1.59226605 1.47400561 1.49290937 1.56632309 1.38951061 1.49821121
 1.53454475 1.52075866 1.51200373 1.3511376 ]
Mean RMSE: 1.4931670676131485


In [48]:
# Step 6: Hyperparameter tuning with GridSearchCV
param_grid = {'regressor__n_estimators': [50, 100, 150],
              'regressor__max_features': [3, 5, 7],
              'regressor__max_depth': [10, 15, 20, None],  # Add max depth to control overfitting
              'regressor__min_samples_split': [2, 5, 10],   # Minimum samples required to split
              'regressor__bootstrap': [False]
}

grid_search = GridSearchCV(pipeline,
                           param_grid,
                           cv=5,
                           scoring='neg_mean_squared_error',
                           return_train_score=True
)

In [49]:
# Fit GridSearchCV
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor',
                                        ColumnTransformer(transformers=[('cat',
                                                                         OneHotEncoder(handle_unknown='ignore',
                                                                                       sparse_output=False),
                                                                         ['COLLISION_TYPE',
                                                                          'WEATHER_CONDITION',
                                                                          'LIGHT_CONDITION',
                                                                          'ROADWAY_SURFACE_COND',
                                                                          'RELATION_TO_ROADWAY',
                                                                          'ROADWAY_ALIGNMENT',
                                                                          'ROADWAY_DEFECT',
                                                                          'ROADWAY_DESCRIPTION',
                                                                          'INTERSECTION_TYPE',
                                                                          'SCHOOL_ZONE...
                                                                          'VDOT_DISTRICT',
                                                                          'AREA_TYPE',
                                                                          'RTE_NM'])])),
                                       ('regressor',
                                        RandomForestRegressor(random_state=42))]),
             param_grid={'regressor__bootstrap': [False],
                         'regressor__max_depth': [10, 15, 20, None],
                         'regressor__max_features': [3, 5, 7],
                         'regressor__min_samples_split': [2, 5, 10],
                         'regressor__n_estimators': [50, 100, 150]},
             return_train_score=True, scoring='neg_mean_squared_error')

In [50]:
# Display the best hyperparameters and best model
print("Best Hyperparameters (GridSearchCV):", grid_search.best_params_)
best_model = grid_search.best_estimator_

Best Hyperparameters (GridSearchCV): {'regressor__bootstrap': False, 'regressor__max_depth': None, 'regressor__max_features': 3, 'regressor__min_samples_split': 10, 'regressor__n_estimators': 150}


In [52]:
# Step 7: Model evaluation on test set
y_test_predictions = best_model.predict(X_test)
final_mse = mean_squared_error(y_test, y_test_predictions)
final_rmse = np.sqrt(final_mse)
print("Final Test RMSE (GridSearchCV):", final_rmse)

# Mean Absolute Error (optional)
final_mae = mean_absolute_error(y_test, y_test_predictions)
print("Final Test MAE (GridSearchCV):", final_mae)

Final Test RMSE (GridSearchCV): 1.3961252376245032
Final Test MAE (GridSearchCV): 1.2602320965608467


In [53]:
# View GridSearchCV results in a DataFrame
results_df = pd.DataFrame(grid_search.cv_results_)
print(results_df[['params', 'mean_test_score', 'std_test_score']].sort_values(by="mean_test_score", ascending=False))

                                               params  mean_test_score  \
89  {'regressor__bootstrap': False, 'regressor__ma...        -1.993977   
97  {'regressor__bootstrap': False, 'regressor__ma...        -1.995220   
98  {'regressor__bootstrap': False, 'regressor__ma...        -1.995404   
88  {'regressor__bootstrap': False, 'regressor__ma...        -1.995735   
96  {'regressor__bootstrap': False, 'regressor__ma...        -1.996614   
..                                                ...              ...   
83  {'regressor__bootstrap': False, 'regressor__ma...        -2.122956   
82  {'regressor__bootstrap': False, 'regressor__ma...        -2.125549   
99  {'regressor__bootstrap': False, 'regressor__ma...        -2.130560   
81  {'regressor__bootstrap': False, 'regressor__ma...        -2.145998   
90  {'regressor__bootstrap': False, 'regressor__ma...        -2.150832   

    std_test_score  
89        0.099733  
97        0.100979  
98        0.098512  
88        0.095441  
96    

In [54]:
# Step 8: Further tuning with RandomizedSearchCV
param_distribs = {'regressor__n_estimators': randint(low=50, high=200),
                  'regressor__max_features': randint(low=3, high=8),
                  'regressor__max_depth': [10, 15, 20, None],
                  'regressor__min_samples_split': [2, 5, 10],
                  'regressor__bootstrap': [False]
}

rnd_search = RandomizedSearchCV(pipeline,
                                param_distributions=param_distribs,
                                n_iter=15,  # Try 15 random combinations
                                cv=5,
                                scoring='neg_mean_squared_error',
                                random_state=42
)

In [55]:
# Fit RandomizedSearchCV
rnd_search.fit(X_train, y_train)

RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('preprocessor',
                                              ColumnTransformer(transformers=[('cat',
                                                                               OneHotEncoder(handle_unknown='ignore',
                                                                                             sparse_output=False),
                                                                               ['COLLISION_TYPE',
                                                                                'WEATHER_CONDITION',
                                                                                'LIGHT_CONDITION',
                                                                                'ROADWAY_SURFACE_COND',
                                                                                'RELATION_TO_ROADWAY',
                                                                                'ROADWAY_ALIGNMENT',
                                                                                'ROADWAY_DEFECT',
                                                                                'ROADWAY_DESCRIPTION',
                                                                                'INTERSECTION_TYPE',
                                                                                'SCHOO...
                   param_distributions={'regressor__bootstrap': [False],
                                        'regressor__max_depth': [10, 15, 20,
                                                                 None],
                                        'regressor__max_features': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x798aa96415a0>,
                                        'regressor__min_samples_split': [2, 5,
                                                                         10],
                                        'regressor__n_estimators': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x798aabd94730>},
                   random_state=42, scoring='neg_mean_squared_error')

In [56]:
# Display the best hyperparameters from RandomizedSearchCV
print("Best Hyperparameters (RandomizedSearchCV):", rnd_search.best_params_)

Best Hyperparameters (RandomizedSearchCV): {'regressor__bootstrap': False, 'regressor__max_depth': None, 'regressor__max_features': 6, 'regressor__min_samples_split': 10, 'regressor__n_estimators': 57}


In [57]:
# Evaluate on test data with the best model from RandomizedSearchCV
y_test_predictions_rnd = rnd_search.best_estimator_.predict(X_test)
final_mse_rnd = mean_squared_error(y_test, y_test_predictions_rnd)
final_rmse_rnd = np.sqrt(final_mse_rnd)
print("Final Test RMSE (RandomizedSearchCV):", final_rmse_rnd)

# Display Mean Absolute Error
final_mae_rnd = mean_absolute_error(y_test, y_test_predictions_rnd)
print("Final Test MAE (RandomizedSearchCV):", final_mae_rnd)

Final Test RMSE (RandomizedSearchCV): 1.4049984264962392
Final Test MAE (RandomizedSearchCV): 1.258694827346143


In [58]:
# Display each combination's performance from RandomizedSearchCV
cvres = rnd_search.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)

1.4185849141428495 {'regressor__bootstrap': False, 'regressor__max_depth': 20, 'regressor__max_features': 6, 'regressor__min_samples_split': 2, 'regressor__n_estimators': 64}
1.4165859961098743 {'regressor__bootstrap': False, 'regressor__max_depth': 20, 'regressor__max_features': 7, 'regressor__min_samples_split': 2, 'regressor__n_estimators': 152}
1.4275382168826232 {'regressor__bootstrap': False, 'regressor__max_depth': 15, 'regressor__max_features': 5, 'regressor__min_samples_split': 10, 'regressor__n_estimators': 124}
1.4193511664585832 {'regressor__bootstrap': False, 'regressor__max_depth': 20, 'regressor__max_features': 7, 'regressor__min_samples_split': 10, 'regressor__n_estimators': 199}
1.4334386458528106 {'regressor__bootstrap': False, 'regressor__max_depth': 10, 'regressor__max_features': 4, 'regressor__min_samples_split': 5, 'regressor__n_estimators': 87}
1.4239825569639375 {'regressor__bootstrap': False, 'regressor__max_depth': 15, 'regressor__max_features': 6, 'regressor_

In [59]:
# Get feature importances from the best model if available
if hasattr(best_model.named_steps['regressor'], 'feature_importances_'):
    feature_importances = best_model.named_steps['regressor'].feature_importances_
    encoded_columns = best_model.named_steps['preprocessor'].named_transformers_['cat'].get_feature_names_out(categorical_columns)
    importance_df = pd.DataFrame(sorted(zip(feature_importances, encoded_columns), reverse=True), columns=["Importance", "Feature"])
    print(importance_df.head(10))  # Display top 10 most important features

   Importance                                            Feature
0    0.013178                             COLLISION_TYPE_12. Ped
1    0.012126          COLLISION_TYPE_9. Fixed Object - Off Road
2    0.011911                        ROADWAY_SURFACE_COND_2. Wet
3    0.011283  ROADWAY_DESCRIPTION_3. Two-Way, Divided, Posit...
4    0.011151        ROADWAY_DESCRIPTION_1. Two-Way, Not Divided
5    0.010915                         COLLISION_TYPE_1. Rear End
6    0.010695                     VDOT_DISTRICT_5. Hampton Roads
7    0.010624                            COLLISION_TYPE_10. Deer
8    0.010613                        LIGHT_CONDITION_2. Daylight
9    0.010547              ROADWAY_ALIGNMENT_1. Straight - Level


# Summary of results:
**Intial cross-validation RMSE Scores (before tuning)**

[1.592, 1.474, 1.492, 1.566, 1.389, 1.498, 1.534, 1.520, 1.512, 1.351]

*  Mean RMSE: 1.493

---



**Best Hyperparamters form GridSearchCV**

{'regressor__bootstrap': False, 'regressor__max_depth': None, 'regressor__max_features': 3, 'regressor__min_samples_split': 10, 'regressor__n_estimators': 150}

---



**Final test RMSE and MAE from GridSearchCV**

* Final Test RMSE: 1.396
* Final Test MAE: 1.260


---

**Best Hyperparameters from RandomizedSearchCV**

{'regressor__bootstrap': False, 'regressor__max_depth': None, 'regressor__max_features': 6, 'regressor__min_samples_split': 10, 'regressor__n_estimators': 57}


---

**Final test RMSE and MAE from RandomizedSearchCV**

*   Final Test RMSE: 1.405
*   Final Test MAE: 1.259


---

 **Feature Importances (Top 10 Features)**

1. Importance: 0.013178 | Feature: COLLISION_TYPE_12. Ped
2. Importance: 0.012126 | Feature: COLLISION_TYPE_9. Fixed Object - Off Road
3. Importance: 0.011911 | Feature: ROADWAY_SURFACE_COND_2. Wet
4. Importance: 0.011283 | Feature: ROADWAY_DESCRIPTION_3. Two-Way, Divided, Positive
5. Importance: 0.011151 | Feature: ROADWAY_DESCRIPTION_1. Two-Way, Not Divided
6. Importance: 0.010915 | Feature: COLLISION_TYPE_1. Rear End
7. Importance: 0.010695 | Feature: VDOT_DISTRICT_5. Hampton Roads
8. Importance: 0.010624 | Feature: COLLISION_TYPE_10. Deer
9. Importance: 0.010613 | Feature: LIGHT_CONDITION_2. Daylight
10. Importance: 0.010547 | Feature: ROADWAY_ALIGNMENT_1. Straight - Level




